### PR TITLE
Add zero_grad(set_to_none=True)

### PR DESCRIPTION
**Pull request by firstprayer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42754 Add reset_grad() function**

Differential Revision: [D23010859](https://our.internmc.facebook.com/intern/diff/D23010859)

### CODE REVIEW

**Comment by firstprayer** [comment]
## Summary
Address issue https://github.com/pytorch/pytorch/issues/41696

## Test Plan:

I leveraged the mnist example in the pytorch/example, manually changed the zero_grad() to reset_grad() and see if that affects the training precision:

Before:
```
Train Epoch: 3 [57600/60000 (96%)]      Loss: 0.049052                                                                                                                                                              │·························································
Train Epoch: 3 [58240/60000 (97%)]      Loss: 0.072383                                                                                                                                                              │·························································
Train Epoch: 3 [58880/60000 (98%)]      Loss: 0.004111                                                                                                                                                              │·························································
Train Epoch: 3 [59520/60000 (99%)]      Loss: 0.001328                                                                                                                                                              │·························································
                                                                                                                                                                                                                    │·························································
Test set: Average loss: 0.0368, Accuracy: 9867/10000 (99%)
```

After:
```
Train Epoch: 3 [58880/60000 (98%)]      Loss: 0.004111                                                                                                                                                              │·························································
Train Epoch: 3 [59520/60000 (99%)]      Loss: 0.001328
Test set: Average loss: 0.0368, Accuracy: 9867/10000 (99%)
```

**Comment by mcarilli** [comment]
Optional suggestion:  https://github.com/pytorch/pytorch/issues/41696#issuecomment-677848472

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -175,6 +176,25 @@ def zero_grad(self):
...
+        2. User can no longer rely on checking if `.grad` is None to see if a tensor
+          is touched in the backward pass
+        3. `nn.optim` optimizers have a different behavior if the gradient is 0 or None
+          (in one case it does the step with a gradient of 0 and in the other it skip
+          the step altogether).
+        """
+        for group in self.param_groups:
+            for p in group['params']:
+                if p.grad is not None:
+                    p.grad.detach_()
```
Comment (mcarilli) [inline torch/optim/optimizer.py:195]: Why do we need to `detach_()`?  Afaict the only way this has an effect is if something external also holds a reference to `.grad` _and_ `.grad` was created in a `create_graph=True` backward pass.

**Comment by vincentqb** [comment]
Thanks for running an example. Out of curiosity, did it change the runtime?

**Comment by vincentqb** [comment]
Responded on API on [original issue](https://github.com/pytorch/pytorch/issues/41696#issuecomment-677911781).

**Comment by mcarilli** [comment]
> Can you please modify `test_zero_grad` in test_nn.py so that it tests both modes?
> @mcarilli, do you think it makes sense to modify grad scaling tests in test_cuda.py to use `.zero_grad` with both options?

Yes, thats a good suggestion. should be several one-liners in the [tests](https://github.com/pytorch/pytorch/blob/c972e6232a0ea1f5993898c26cd4f605c1a60cd8/test/test_cuda.py#L1979) after this PR is merged.

File: torch/optim/optimizer.py (hunk 1 above)
Comment (mcarilli) [reply torch/optim/optimizer.py]: My point is `detach_` may be the _unsafe_ option.  Because it affects `p.grad` in place, it silently affects anything else holding a reference to `p.grad`.  Setting `p.grad = None` without `detach_` simply drops our reference.  Anything else holding a reference to `grad` will not see an effect.

Admittedly, I'm not sure why the default zeroing behavior of `zero_grad` performs `detach_` then `zero_`.  I assume it's to avoid building up spurious autograd history if the grad was created with `create_graph=True` and therefore requires grad.

If so, in the alternative set-to-None path, we dont need `detach_`.  We drop the reference to `.grad`, we don't perform any ops on it, therefore there's no danger of building spurious autograd history.  And if we don't `detach_`, we don't risk silently affecting other references to `grad`.

tl;dr I think _not_ detaching is the safe option here.

@albanD you're good with these tricky cases...Why is `detach_` used with the default zeroing behavior of `zero_grad`?  Also, do you agree _not_ detaching is the safer implementation for the set-to-None path?

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1324,11 +1328,15 @@ def zero_grad(self) -> None:
 
         for p in self.parameters():
             if p.grad is not None:
-                if p.grad.grad_fn is not None:
+                if set_to_none:
                     p.grad.detach_()
```
Comment (albanD) [inline torch/nn/modules/module.py:1332]: I think you can remove this `detach()` as you remove it anyways.

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -1313,8 +1313,12 @@ def requires_grad_(self: T, requires_grad: bool = True) -> T:
             p.requires_grad_(requires_grad)
         return self
 
-    def zero_grad(self) -> None:
-        r"""Sets gradients of all model parameters to zero."""
+    def zero_grad(self, set_to_none=False) -> None:
+        r"""Sets gradients of all model parameters to zero.
+
+        Arguments:
+            set_to_none (bool): instead of setting to zero, set the grad to None.
```
Comment (albanD) [inline torch/nn/modules/module.py:1320]: Can you link to the `nn.optim` version of this doc that contains more details about the change of behavior?

File: torch/optim/optimizer.py [hunk 4]
```diff
@@ -164,16 +164,33 @@ def update_group(group, new_group):
...
+                3. `nn.optim` optimizers have a different behavior if the gradient is 0 or None
+                (in one case it does the step with a gradient of 0 and in the other it skip
+                the step altogether).
+        """
         for group in self.param_groups:
             for p in group['params']:
                 if p.grad is not None:
-                    if p.grad.grad_fn is not None:
+                    if set_to_none:
                         p.grad.detach_()
```
Comment (albanD) [inline torch/optim/optimizer.py:186]: unneeded detach here as well.

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -1313,8 +1313,12 @@ def requires_grad_(self: T, requires_grad: bool = True) -> T:
             p.requires_grad_(requires_grad)
         return self
 
-    def zero_grad(self) -> None:
-        r"""Sets gradients of all model parameters to zero."""
+    def zero_grad(self, set_to_none=False) -> None:
```
Comment (albanD) [inline torch/nn/modules/module.py:1316]: nit: can you add the type for `set_to_none`: `set_to_none: bool = False`

File: torch/optim/optimizer.py [hunk 6]
```diff
@@ -164,16 +164,33 @@ def update_group(group, new_group):
             update_group(g, ng) for g, ng in zip(groups, saved_groups)]
         self.__setstate__({'state': state, 'param_groups': param_groups})
 
-    def zero_grad(self):
-        r"""Clears the gradients of all optimized :class:`torch.Tensor` s."""
+    def zero_grad(self, set_to_none=False):
```
Comment (albanD) [inline torch/optim/optimizer.py:167]: Can you update `optimizer.pyi` as well to reflect this change? Just add the new arg and type there.

File: torch/optim/optimizer.py [hunk 7]
```diff
@@ -164,16 +164,33 @@ def update_group(group, new_group):
...
+
+        Arguments:
+            set_to_none (bool): instead of setting to zero, set the grad to None.
+                This will in general have lower memory footprint, but using this
+                comes with caveats, to name a few:
+                1. When user tries to access the gradient value and perform manual ops on it.
+                A None attribute or a Tensor full of 0s will be different.
+                2. User can no longer rely on checking if `.grad` is None to see if a tensor
+                is touched in the backward pass
+                3. `nn.optim` optimizers have a different behavior if the gradient is 0 or None
```
Comment (mcarilli) [inline torch/optim/optimizer.py:178]: `torch.optim`

File: torch/optim/optimizer.py (hunk 1 above)
Comment (albanD) [reply torch/optim/optimizer.py]: The original code is trying to make sure that we don't change the Tensor referenced by `.grad` as much as possible I think.
And yes as mentioned below, I do agree that we don't want to detach here.

File: torch/optim/optimizer.py [hunk 8]
```diff
@@ -164,16 +164,33 @@ def update_group(group, new_group):
...
 
-    def zero_grad(self):
-        r"""Clears the gradients of all optimized :class:`torch.Tensor` s."""
+    def zero_grad(self, set_to_none=False):
+        r"""Set the gradients of all optimized :class:`torch.Tensor` s to zero.
+
+        Arguments:
+            set_to_none (bool): instead of setting to zero, set the grad to None.
+                This will in general have lower memory footprint, but using this
+                comes with caveats, to name a few:
```
Comment (mcarilli) [inline torch/optim/optimizer.py:173]: I wouldn't say these are caveats.  Some are benefits, like point 3 (skipping the update is faster if the grad is None).
Instead of
```
This will in general have lower memory footprint, but using this comes with caveats, to name a few:
```
I would say
```
This is will in general have lower memory footprint, and can modestly improve performance.  However, it changes certain behaviors.  For example:
```

File: torch/nn/modules/module.py (hunk 3 above)
Comment (mcarilli) [reply torch/nn/modules/module.py:1320]: `torch.optim`, right?

File: torch/optim/optimizer.py [hunk 9]
```diff
@@ -164,16 +164,33 @@ def update_group(group, new_group):
...
+                1. When user tries to access the gradient value and perform manual ops on it.
+                A None attribute or a Tensor full of 0s will be different.
+                2. User can no longer rely on checking if `.grad` is None to see if a tensor
```
Comment (mcarilli) [inline torch/optim/optimizer.py:176]: this doesn't make sense.  I'd say it's the opposite:  if a user sets grads = None before backward, they CAN check if `.grad` is None to see if the tensor received a gradient in the backward pass.  Let's just describe the behavior:
```
If the user requests `zero_grad(set_to_none=True)` followed by a backward pass, `.grad`\ s are guaranteed to be None for params that did not receive a gradient.
```

File: torch/nn/modules/module.py (hunk 3 above)
Comment (albanD) [reply torch/nn/modules/module.py:1321]: Yes, typo!

File: torch/optim/optimizer.py [hunk 10]
```diff
@@ -164,16 +164,32 @@ def update_group(group, new_group):
...
         self.__setstate__({'state': state, 'param_groups': param_groups})
 
-    def zero_grad(self):
-        r"""Clears the gradients of all optimized :class:`torch.Tensor` s."""
+    def zero_grad(self, set_to_none: bool = False):
+        r"""Set the gradients of all optimized :class:`torch.Tensor` s to zero.
+
+        Arguments:
+            set_to_none (bool): instead of setting to zero, set the grad to None.
+                This is will in general have lower memory footprint, and can modestly improve performance.
```
Comment (mcarilli) [inline torch/optim/optimizer.py:172]: `This is will` -> `This will`

File: torch/optim/optimizer.py [hunk 11]
```diff
@@ -164,16 +164,32 @@ def update_group(group, new_group):
...
+                However, it changes certain behaviors. For example:
+                1. When user tries to access the gradient value and perform manual ops on it.
+                A None attribute or a Tensor full of 0s will be different.
```
Comment (mcarilli) [inline torch/optim/optimizer.py:175]: combine sentences
```
When user tries to access the gradient value and perform manual ops on it,
                a None attribute or a Tensor full of 0s will act differently.
```

File: torch/optim/optimizer.py [hunk 12]
```diff
@@ -164,16 +164,32 @@ def update_group(group, new_group):
...
+                2. If the user requests `zero_grad(set_to_none=True)` followed by a backward pass, `.grad` s
```
Comment (mcarilli) [inline torch/optim/optimizer.py:176]: Upon rereading this, I think we should switch the order of points 1 and 2.  The current point 2 sets up some context for (what is currently) point 1, so I think it makes sense to say point 2 first.

**Comment by ngimel** [comment]
@alband, @vincentqb anything holding up this diff? Comments seem to be addressed (other than @mcarilli's doc suggestions which are imo minor).

File: torch/nn/modules/module.py (hunk 3 above)
Comment (mcarilli) [reply torch/nn/modules/module.py:1321]:
```
See :func:`torch.optim.optimizer.zero_grad` for details.
```

**Comment by albanD** [comment]
Nothing blocking on my side. Just the doc update. And make sure that the CI is happy with it after rebase.

**Review by ngimel — APPROVED** [review]
Approving, subject to minor doc fixes.

**Review by vincentqb — APPROVED** [review]
LGTM too, I fixed two of the three doc changes that I see. I'm ok either way with the last one:
> Upon rereading this, I think we should switch the order of points 1 and 2. The current point 2 sets up some context for (what is currently) point 1, so I think it makes sense to say point 2 first.

**Review by mcarilli — CHANGES_REQUESTED** [review]
https://github.com/pytorch/pytorch/pull/42754#pullrequestreview-481216847 is not a big deal imo but don't forget https://github.com/pytorch/pytorch/pull/42754/files#r481400744.
